File: client/src/lib/sacrificeCost.ts
// sacrificeCost.ts — #747: the client half of a sacrifice cost of N
// permanents ("Sacrifice two artifacts", "Sacrifice five Treasures").
//
// There is no count field on the wire. A sacrifice clause ships as
// `sacrifice_options` (a LegalTargetsView) on an activated ability, a
// mana ability, or a hand card's additional cost, and its `min` /
// `max` ARE the count — always equal, 1 for "Sacrifice a creature".
// The server ships the options in payment order (tokens first, then
// lower mana value, then the ability's own source, then board order),
// which is the order the legal-move enumerator takes its one payment
// from; "Choose for me" takes the first N of that list, so the button
// picks exactly what a bot would. See docs/protocol.md, "Sacrifice
// costs of N permanents".
//
// Pure functions only, so the picker and the menus share one
// definition of "how many" and "can this be paid" and vitest can pin
// both without mounting a component.

import type { CardView } from "./protocol";

// SacrificeOptionsShape is the part of a LegalTargetsView the helpers
// read. The menus' local cost shapes declare only cards / players, so
// min / max are optional here.
export interface SacrificeOptionsShape {
  cards?: string[];
  min?: number;
  max?: number;
  // #1213: the clause's count is the announced X ('Sacrifice X
  // Treasures'), so min / max say nothing and the number picked IS
  // the x_value the same message announces. See sacrificeRange.
  count_from_x?: boolean;
}

// sacrificeCount is how many permanents the clause sacrifices: the
// view's max, and never less than one — a clause that reached the
// client with no count (an older server, a hand-built fixture) is the
// "Sacrifice a creature" every cost was before #747.
export function sacrificeCount(opts: SacrificeOptionsShape | undefined): number {
  const n = opts?.max ?? opts?.min ?? 1;
  return n >= 1 ? n : 1;
}

// sacrificeShortfall is the reason a sacrifice cost can't be paid
// right now, or "" when it can. The one-permanent wording is unchanged
// from before #747; a count names what is missing: "needs three Foods
// (you have 2)".
export function sacrificeShortfall(opts: SacrificeOptionsShape | undefined, label: string): string {
  if (!opts) return "";
  const have = opts.cards?.length ?? 0;
  const need = sacrificeCount(opts);
  if (have >= need) return "";
  if (need === 1) return `nothing to sacrifice (${label})`;
  return `needs ${label} (you have ${have})`;
}

// orderSacrificeOptions resolves the server's option IDs against the
// cards the client holds, KEEPING THE SERVER'S ORDER — the picker lists
// them in payment order, so the "Choose for me" picks are the top of
// the list. IDs with no matching card are dropped.
export function orderSacrificeOptions(board: CardView[], ids: string[] | undefined): CardView[] {
  if (!ids || ids.length === 0) return [];
  const byID = new Map(board.map((c) => [c.instance_id, c]));
  const out: CardView[] = [];
  for (const id of ids) {
    const c = byID.get(id);
    if (c) out.push(c);
  }
  return out;
}

// toggleSacrificePick is one click in the picker. At a count of 1 a
// click replaces the pick, as the single-choice picker always did. At
// a count of N a click adds an unpicked permanent (unless N are
// already picked) or removes a picked one.
export function toggleSacrificePick(chosen: string[], id: string, count: number): string[] {
  if (count <= 1) return chosen.length === 1 && chosen[0] === id ? chosen : [id];
  if (chosen.includes(id)) return chosen.filter((c) => c !== id);
  if (chosen.length >= count) return chosen;
  return [...chosen, id];
}

// chooseForMeState is whether the picker shows the "Choose for me"
// button, and whether it can be pressed: shown only for a count of two
// or more (at one, a single click is already the whole choice), and
// disabled when fewer options than the count are on offer.
export function chooseForMeState(
  count: number,
  optionCount: number,
): { shown: boolean; disabled: boolean } {
  const shown = count > 1;
  return { shown, disabled: shown && optionCount < count };
}

// chooseSacrificeForMe fills the selection with the first N options in
// the order given — the server's payment order. It only SELECTS; the
// player still confirms, and can change the picks first.
export function chooseSacrificeForMe(options: string[], count: number): string[] {
  return options.slice(0, Math.max(0, count));
}

// canConfirmSacrifice is the confirm gate: exactly N distinct picks.
export function canConfirmSacrifice(chosen: string[], count: number): boolean {
  return chosen.length === count && new Set(chosen).size === count;
}

// keepAvailablePicks drops picks that are no longer among the options:
// a chosen Treasure destroyed in response while the picker is open
// would otherwise stay in the selection unseen, unable to be unticked,
// holding a slot that disables every other row, and a confirm would
// send an ID the server refuses. Returns `chosen` itself when nothing
// was dropped, so a reactive caller can compare by reference.
export function keepAvailablePicks(chosen: string[], optionIDs: string[]): string[] {
  const available = new Set(optionIDs);
  const kept = chosen.filter((id) => available.has(id));
  return kept.length === chosen.length ? chosen : kept;
}

// --- #1213: a count the ACTIVATOR announces --------------------------
//
// Two printed clauses no longer ship min == max:
//
//   "Sacrifice one or more artifacts"  min 1, max 0 (no ceiling)
//   "Sacrifice X Treasures"            count_from_x, the count IS the X
//
// Everything above stays the fixed-count vocabulary, byte for byte,
// because every other cost site still uses it. What follows is the
// RANGE vocabulary, and a fixed clause passes through it unchanged:
// its bounds are N..N and every answer these give is the answer the
// functions above give.

// SacrificeRange is how many permanents one payment may name. `max`
// of 0 means "no printed ceiling" — the board is the only bound — and
// a caller that needs a concrete number substitutes the option count.
export interface SacrificeRange {
  min: number;
  max: number;
}

// sacrificeRange reads the bounds off the view. A clause whose count
// is the announced X (`count_from_x`) has no printed bounds at all, so
// it reads as "at least one, as many as you control": the client sends
// the number picked AS the x_value, which is the only reading that
// cannot disagree with the server.
//
// A view with no bounds at all is the one-permanent clause every cost
// was before #747, which is what sacrificeCount already assumes.
export function sacrificeRange(opts: SacrificeOptionsShape | undefined): SacrificeRange {
  if (opts?.count_from_x) return { min: 1, max: 0 };
  const min = opts?.min ?? 0;
  const max = opts?.max ?? 0;
  if (min <= 0 && max <= 0) return { min: 1, max: 1 };
  if (min <= 0) return { min: max, max };
  return { min, max };
}

// sacrificeCeiling is the range's `max` resolved against what is
// actually on offer: an unbounded clause is capped by the board.
export function sacrificeCeiling(range: SacrificeRange, optionCount: number): number {
  return range.max > 0 ? range.max : optionCount;
}

// canConfirmSacrificeRange is the confirm gate for a range: at least
// `min` distinct picks, and no more than the ceiling.
export function canConfirmSacrificeRange(
  chosen: string[],
  range: SacrificeRange,
  optionCount: number,
): boolean {
  if (new Set(chosen).size !== chosen.length) return false;
  return chosen.length >= range.min && chosen.length <= sacrificeCeiling(range, optionCount);
}

// toggleSacrificePickInRange is one click in a ranged picker. A
// single-pick clause (ceiling 1) replaces the pick, exactly as
// toggleSacrificePick does; anything wider adds until the ceiling and
// removes on a second click.
export function toggleSacrificePickInRange(
  chosen: string[],
  id: string,
  ceiling: number,
): string[] {
  return toggleSacrificePick(chosen, id, ceiling);
}

// sacrificeRangeShortfall is the reason a ranged cost cannot be paid
// right now, or "" when it can — the greyed-row text, one clause over
// from sacrificeShortfall. It reads the FLOOR, because a clause with
// no ceiling is payable the moment one permanent can pay it.
export function sacrificeRangeShortfall(
  opts: SacrificeOptionsShape | undefined,
  label: string,
): string {
  if (!opts) return "";
  const have = opts.cards?.length ?? 0;
  const need = sacrificeRange(opts).min;
  if (have >= need) return "";
  if (need === 1) return `nothing to sacrifice (${label})`;
  return `needs ${label} (you have ${have})`;
}

File: server/internal/game/sacrifice_cost.go
package game

import (
	"sort"

	"github.com/google/uuid"
)

// sacrifice_cost.go — #747 (ADR 0020 addendum §12–§15, ADR 0021
// addendum): a sacrifice cost of N permanents. "Sacrifice two
// artifacts" (Sai), "Sacrifice three Foods" (Samwise Gamgee),
// "Sacrifice five Treasures" (Magda, Brazen Outlaw).
//
// There is no new cost component. The count lives on the sacrifice
// clause itself, as the TargetSpec's Min == Max, so the three cost
// sites that already carry the clause — AbilityCost.SacrificeOther,
// ManaAbilityShape.SacrificeOther and AdditionalCost.Sacrifice — get
// the count with no new field, and a cost merge (effects.Plus) cannot
// separate the count from the thing it counts.
//
// Three pieces live here because all three sites and both readers
// (the legal enumerator and the protocol view) share them:
//
//   - SacrificeCostCount reads the count off a clause.
//   - payCostSacrificesLocked pays one payment as one simultaneous
//     exit.
//   - SacrificePaymentOrderForEffect is the policy-neutral order the
//     enumerator takes its one payment from, and the order the view
//     ships the options in, so the client's "Choose for me" fills the
//     picker with the same permanents a bot would pay.

// SacrificeCostCount is how many permanents a FIXED sacrifice clause
// pays: the spec's Max. A hand-built fixture clause that never set a
// count is read as 1, the count every sacrifice cost had before #747.
// Nil-safe: a nil clause pays nothing.
//
// It answers for a fixed clause only. A VARIABLE one (#1213 —
// "sacrifice one or more", "sacrifice X") has no single number to
// give, and reads here as its floor, which is the weaker answer and
// never the one a payment should be built from. Ask
// SacrificeCostBounds instead, and SacrificeCostVariable first if you
// need to know which kind you have.
func SacrificeCostCount(spec *TargetSpec) int {
	if spec == nil {
		return 0
	}
	if spec.CountFromX {
		return 0
	}
	if spec.Max < 1 {
		if spec.Min > 1 {
			return spec.Min
		}
		return 1
	}
	return spec.Max
}

// SacrificeCostVariable reports a clause whose count the ACTIVATOR
// announces rather than one the card prints (#1213):
//
//   - "Sacrifice one or more artifacts" (Radiant Lotus) — a floor of
//     one with no printed ceiling, Min ≥ 1 and Max 0.
//   - "Sacrifice X Treasures" (Grim Hireling) — CountFromX, the count
//     announced at CR 602.2b with the rest of the activation.
//
// Nil-safe (false). The readers that care are the announce path (how
// many IDs is this payment allowed to name), the protocol view (what
// min / max does the picker enforce) and the legal enumerator (how
// many payments does it offer).
func SacrificeCostVariable(spec *TargetSpec) bool {
	if spec == nil {
		return false
	}
	return spec.CountFromX || spec.Max != spec.Min
}

// SacrificeCountFromX reports the "Sacrifice X …" form specifically —
// the one whose count is the announced X rather than a number the
// activator is free to pick. Nil-safe (false).
//
// It is what makes AbilityCost.DemandsX() true for Grim Hireling,
// whose mana component ({B}) has no {X} slot at all.
func SacrificeCountFromX(spec *TargetSpec) bool {
	return spec != nil && spec.CountFromX
}

// SacrificeCostBounds is how many permanents ONE payment of a
// sacrifice clause may name, given the X announced at CR 601.2b /
// 602.2b (#1213). `lo` is the floor and `hi` the ceiling; **hi == 0
// means there is no printed ceiling** and the activator's own board is
// the only bound.
//
// The three shapes, and the whole of the variable-count design:
//
//	fixed        Min == Max == N        lo = hi = N
//	one or more  Min >= 1, Max == 0     lo = Min, hi = 0   (open)
//	X            CountFromX             lo = hi = x
//
// A CountFromX clause with a negative announced X reads as zero rather
// than as a negative bound: the announce path rejects a negative X
// separately, and a bound that could go negative would make an empty
// payment look illegal for the wrong reason.
//
// Nil-safe: a nil clause pays nothing, 0 / 0 — which a caller must
// read as "no component", not as "an open count", so every caller
// checks the clause for nil before asking.
func SacrificeCostBounds(spec *TargetSpec, x int) (lo, hi int) {
	if spec == nil {
		return 0, 0
	}
	if spec.CountFromX {
		if x < 0 {
			x = 0
		}
		return x, x
	}
	n := SacrificeCostCount(spec)
	if spec.Max < 1 && spec.Min >= 1 {
		// "One or more": a floor the card prints and no ceiling.
		return spec.Min, 0
	}
	return n, n
}

// SacrificeCountLegal reports whether naming `named` permanents is a
// legal payment of `spec` at the announced X — the one predicate the
// validator, the protocol view and the legal enumerator share, so
// none of them can offer or accept a count the others refuse (#544).
//
// The CountFromX form is answered FIRST rather than through the
// bounds, because the two readings of `hi == 0` collide there: it is
// the "no printed ceiling" sentinel for an open count, and it is the
// real and only legal answer for an X announced as zero. A clause
// whose count IS the announced X has one legal payment and this says
// so directly.
//
// Nil-safe: a nil clause is paid by naming nothing.
func SacrificeCountLegal(spec *TargetSpec, x, named int) bool {
	if spec == nil {
		return named == 0
	}
	if spec.CountFromX {
		if x < 0 {
			x = 0
		}
		return named == x
	}
	lo, hi := SacrificeCostBounds(spec, x)
	if named < lo {
		return false
	}
	return hi == 0 || named <= hi
}

// payCostSacrificesLocked sacrifices every permanent of one cost
// payment — the source when the cost sacrifices it, and the N
// permanents of the clause — as ONE simultaneous battlefield exit.
//
// Each permanent still goes through sacrificePermanentLocked, so each
// emits its own EventSacrifice and takes its own route to the
// graveyard: "whenever you sacrifice a permanent" and dies triggers
// fire once per permanent. What the batch adds is what the watchers
// see. Leaves-the-battlefield and sacrifice triggers look back in time
// (CR 603.10a), so a Blood Artist sacrificed alongside a Goblin must
// see both deaths. Paid one at a time with no batch, the Artist would
// see the Goblin only when the Goblin happened to go first, and the
// trigger count would depend on the order of IDs on the wire.
//
// beginSimultaneousExitLocked has no indestructible filter (that lives
// in DestroyPermanentsForEffect, which this does not call), so an
// indestructible permanent is still sacrificed, as CR 701.21a says.
//
// Caller must hold g.mu in write mode, and must have validated the
// payment with validateSacrificeCostLocked.
func (g *Game) payCostSacrificesLocked(ids []uuid.UUID, answers map[uuid.UUID]bool) error {
	if len(ids) == 0 {
		return nil
	}
	defer g.beginSimultaneousExitLocked(ids)()
	for _, id := range ids {
		// #1397: `answers` are the CR 903.9 answers the commanders'
		// owners gave before the payment began; nil for a payer that
		// asks nobody (the auto-tapper), which is "unasked".
		if err := g.sacrificeAnsweredLocked(id, commanderAnswerFor(answers, id)); err != nil {
			return err
		}
	}
	return nil
}

// SacrificePaymentOrderForEffect returns `ids` — candidates for a
// sacrifice cost — in the order a payment takes them from (ADR 0020
// addendum §15):
//
//  1. tokens before nontokens;
//  2. then lower mana value;
//  3. then the ability's own source last, when it is a candidate
//     (pass uuid.Nil when there is no source, as for a spell's
//     additional cost);
//  4. then the order given, which callers pass in board order, so the
//     result is stable.
//
// The order is policy-neutral on purpose. The legal enumerator offers
// one sacrifice-N payment per move — the first N of this order —
// rather than every combination, and `legal` must not import a bot
// policy (#687). The protocol view ships its options in the same
// order, so the owner's "Choose for me" button (#747) fills the picker
// with exactly the permanents a bot would have paid.
//
// A new slice; `ids` is not modified. IDs not on the battlefield sort
// as nontokens of mana value 0, which no caller passes.
//
// Caller must hold g.mu (read or write).
func (g *Game) SacrificePaymentOrderForEffect(ids []uuid.UUID, sourceID uuid.UUID) []uuid.UUID {
	type key struct {
		id     uuid.UUID
		token  bool
		mv     int
		source bool
	}
	keys := make([]key, 0, len(ids))
	for _, id := range ids {
		k := key{id: id, source: sourceID != uuid.Nil && id == sourceID}
		if c := findBattlefieldCard(g, id); c != nil {
			k.token = typeLineHas(c.TypeLine, "token")
			k.mv = c.ManaValue()
		}
		keys = append(keys, k)
	}
	sort.SliceStable(keys, func(i, j int) bool {
		a, b := keys[i], keys[j]
		if a.token != b.token {
			return a.token
		}
		if a.mv != b.mv {
			return a.mv < b.mv
		}
		if a.source != b.source {
			return b.source
		}
		return false
	})
	out := make([]uuid.UUID, len(keys))
	for i, k := range keys {
		out[i] = k.id
	}
	return out
}
